pebble/pebble.go

```diff
@@ -78,6 +79,83 @@ func (s *PebbleDHStore) MergeIndexes(indexes []dhstore.Index) error {
 	return batch.Commit(pebble.NoSync)
 }
 
+// DeleteIndexes removes dh-multihash to encrypted-valueKey mappings. This is
+// the inverse of MergeIndexes.
+func (s *PebbleDHStore) DeleteIndexes(indexes []dhstore.Index) error {
+	keygen := s.p.leaseSimpleKeyer()
+	defer keygen.Close()
+	batch := s.db.NewBatch()
+
+	for _, index := range indexes {
+		dmh, err := multihash.Decode(index.Key)
+		if err != nil {
+			return dhstore.ErrMultihashDecode{Err: err, Mh: index.Key}
+		}
+		if multicodec.Code(dmh.Code) != multicodec.DblSha2_256 {
+			return dhstore.ErrUnsupportedMulticodecCode{Code: multicodec.Code(dmh.Code)}
+		}
+
+		// Lookup the encrypted multihash keys for this dh-multihash.
+		mhk, err := keygen.multihashKey(index.Key)
+		if err != nil {
+			return err
+		}
+		vkb, vkbClose, err := s.db.Get(mhk.buf)
+		if err != nil {
+			_ = mhk.Close()
+			if errors.Is(err, pebble.ErrNotFound) {
+				continue
+			}
+			return err
+		}
+		encValueKeys, err := s.unmarshalEncryptedIndexKeys(vkb)
+		vkbClose.Close()
+		if err != nil {
+			_ = mhk.Close()
+			return err
+		}
+
+		// Remove the encrypted value key from the returned set of values.
+		var removed bool
+		for i, evk := range encValueKeys {
+			if bytes.Equal(evk, index.Value) {
+				if len(encValueKeys) == 1 {
+					encValueKeys = nil
+				} else {
+					encValueKeys[i] = encValueKeys[len(encValueKeys)-1]
+					encValueKeys = encValueKeys[:len(encValueKeys)-1]
+				}
+				removed = true
+				break
+			}
+		}
+		if len(encValueKeys) == 0 {
+			if err = batch.Delete(mhk.buf, pebble.NoSync); err != nil {
```

Comment:
I have a suspicion that this would be slow at scale. But it is certainly a good first pass in implementing deletion 👍 

Before doing it any other way, I recommend adding benchmarks. We can then iterate on alternative approaches.

----

On alternative approaches:

I have thought of 2 other ways of doing this, which can be mixed and matched for a high performance opportunistic deletion mechanism:

1. Accumulate indexes marked for deletion, sort them then delete in much larger balks using _delete range_.
2. Update the merger implementation to intelligently recognise entires pending deletion and use merging function to exclude them from the merge, which will effectively result in deletion. This is the same approach I originally [implemented](https://github.com/ipni/go-indexer-core/blob/main/store/pebble/vk_merger.go#L20) in the non-encrypted pebble for opportunistic deletion.

Reply:
This is at least a batched delete (`batch.Delete` and not `s.db.Delete`) so all the deletions are collected in a batch and done in one shot and it should be faster than a large number of individual calls to `s.db.Delete`.

Since this is done by GC, and is not in the critical path of indexing or lookup, it is OK for it to be slow. GC can crawl along in the background at a very slow pace, as long as it does not interfere with indexing or lookup, and it is better to have a slow GC than to do something that requires additional work to be done in the indexing or lookup path. 

The suggested approaches are good in general, but I think that since deletion will be done by GC I would suggest that we not do 2 since that requires work in the ingestion path. Number 1 may be more efficient, but if it only improves GC then the efficiency does not matter at the cost of storing temporary data.

Reply:
I suspect that the compaction cost would be lower with range deletion. For approach No. 2 i think all the changes would remain in this repo if by ingestion path changes we mean changes in storetheindex? 

Thank you for implementing the batch deletion. That should help us gather some data to see if it needs optimisation at all 👍